### PR TITLE
Add support for Kinesis Data Streams Enhanced Fan-out (Consumers)

### LINKED ISSUE #5510: Add support for AWS Kinesis enhanced stream consumer for Lambda
# This is a Feature Proposal

## Description

* Provide support for configuring AWS enhanced fan out Kinesis stream consumers for Lambda that was released:

https://aws.amazon.com/about-aws/whats-new/2018/11/aws-lambda-supports-kinesis-data-streams-enhanced-fan-out-and-http2/
https://docs.aws.amazon.com/lambda/latest/dg/with-kinesis.html#services-kinesis-configure

* The configuration takes streams Event Source options:
--Kinesis Stream
--Consumer (this is what needs to be added)
--Batch Size
--Starting Position
--Enabled

Related Serverless Docs:
https://serverless.com/framework/docs/providers/aws/events/streams/

Similar or dependent issues:

### PR DESCRIPTION
**Pull request by zaccharles**
## What did you implement

Closes #5510

I added the a new configuration property for Kinesis data stream events which configures the event source mapping to use a stream consumer instead of shard iterator.

It creates a new stream consumer when set to `true` and uses an existing consumer if given the ARN of one.

It manages the necessary IAM permissions and supports functions such as `Ref` and `ImportValue` (essential since stream consumer ARNs are AWS generated with timestamps in them).

This is my first PR of this size to this project, so apologies if I didn't quite get some conventions right.

## How can we verify it

```yaml
service: dct

provider:
  name: aws
  runtime: nodejs12.x

functions:
  hello:
    handler: handler.hello
    events:
      - stream:
          batchSize: 1
          type: kinesis
          arn:
            Fn::GetAtt:
              - ExampleStream
              - Arn
          consumer: true
      - stream:
          batchSize: 1
          type: kinesis
          arn:
            Fn::GetAtt:
              - ExampleStream
              - Arn
          consumer:
            Ref: ExampleStreamConsumer

resources:
  Resources:
    ExampleStream: 
      Type: AWS::Kinesis::Stream 
      Properties: 
          ShardCount: 1
    ExampleStreamConsumer: 
      Type: AWS::Kinesis::StreamConsumer
      Properties: 
          StreamARN:
            Fn::GetAtt:
              - ExampleStream
              - Arn
          ConsumerName: whatever
```

## Todos

No todos at the moment.

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 1]
```diff
@@ -134,6 +157,49 @@ class AwsCompileStreamEvents {
               if (typeof event.stream.enabled !== 'undefined') {
                 Enabled = event.stream.enabled ? 'True' : 'False';
               }
+              if (typeof event.stream.consumer !== 'undefined') {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: Let's just write it as:`if (event.stream.consumer)` (I know that in other cases it's `typeof event.stream.x !== 'undefined'`, but it's bad error-prone pattern, let's not repeat that)

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 2]
```diff
@@ -134,6 +157,49 @@ class AwsCompileStreamEvents {
...
+                if (typeof event.stream.consumer === 'boolean') {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: `if (event.stream.consumer === true)`

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 3]
```diff
@@ -134,6 +157,49 @@ class AwsCompileStreamEvents {
...
+                      this.resolveInvalidKinesisConsumerErrorMessage(functionName)
+                    );
+                  }
+                } else if (typeof event.stream.consumer === 'string') {
+                  if (event.stream.consumer.indexOf('arn:') < 0) {
+                    throw new this.serverless.classes.Error(
+                      this.resolveInvalidKinesisConsumerErrorMessage(functionName)
+                    );
+                  }
+                } else if (typeof event.stream.consumer === 'object') {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: `} else {`

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 4]
```diff
@@ -327,6 +397,42 @@ class AwsCompileStreamEvents {
               [streamLogicalId]: streamResource,
             };
 
+            if (event.stream.consumer) {
+              if (streamType === 'kinesis') {
+                if (typeof event.stream.consumer === 'boolean') {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: `if (event.stream.consumer === true)`

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 5]
```diff
@@ -134,6 +157,49 @@ class AwsCompileStreamEvents {
...
+                    )
+                  ) {
+                    const errorMessage = [
+                      `'${JSON.stringify(
+                        event.stream
+                      )}'Bad dynamic property on stream event in function "${functionName}"`,
+                      ' If you use a dynamic "consumer" (such as with Fn::GetAtt, Fn::Join, Ref',
+                      ' or Fn::ImportValue) there must only be one key (either Fn::GetAtt, Fn::Join, Ref',
+                      ' or Fn::ImportValue) in the consumer object. Please check the docs for more info.',
+                    ].join('');
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: I believe this validation is already made on AWS side. As we just pass through this value (and not try to interpret it for our needs) I would leave it without local validation here.

File: lib/plugins/aws/package/compile/events/stream/index.test.js [hunk 6]
```diff
@@ -37,6 +37,81 @@ describe('AwsCompileStreamEvents', () => {
   });
 
   describe('#compileStreamEvents()', () => {
+    it('should throw an error if the "consumer" property is not a string or a boolean', () => {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.test.js]: It's fine if such errors surface with AWS CF validation, we don't need to double validate them here (doing so, we may introduce other issues, and prevent well configured setup from working)

File: lib/plugins/aws/package/compile/events/stream/index.js (hunk 1 above)
Comment (zaccharles) [reply lib/plugins/aws/package/compile/events/stream/index.js]: Okay, sure.

**Comment by zaccharles** [comment]
@medikoo Hi, thanks for the review. I've made some more changes. Specifically, I've relied more on CloudFormation validating things, which is probably a good approach. This has allowed me to simplify things a bit. I'm happy to accept more feedback.

File: lib/plugins/aws/package/compile/events/stream/index.js [hunk 7]
```diff
@@ -134,6 +149,17 @@ class AwsCompileStreamEvents {
               if (typeof event.stream.enabled !== 'undefined') {
                 Enabled = event.stream.enabled ? 'True' : 'False';
               }
+              if (
+                event.stream.consumer &&
+                ['boolean', 'string', 'object'].indexOf(typeof event.stream.consumer) < 0
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/stream/index.js]: Let's skip such validation. In JS world, more natural handling would be as:

```javascript
if (event.stream.consumer) {
  // There's a consumer config
  if (event.stream.consumer === true) {
    // 'true' case
  } else {
    // object case
    if (!event.stream.consumer.requiredProperty) {
      throw new ServerlessError("Missing ...");
    }
  }
}
```

We'll also have schema  based validation brought with: https://github.com/serverless/serverless/issues/6562 so incorporating sophisticated type checks into logic now will add unnecessary maintenance burden.

--- 

So summarizing, I think we should not do any validation here, as _object case_ should be validated by AWS directly, while other cased are picked in straightforward way.

Comment (zaccharles) [reply lib/plugins/aws/package/compile/events/stream/index.js]: Oh okay. I don't write JS every day so I appreciate these tips :)

#6562 looks great!

**Comment by zaccharles** [comment]
It's quite a lot simpler now without any of the validation 😄

**Comment by xharsin** [comment]
@zaccharles Is it possible to use the function name defined with name in serverless under function. 
```yml
 functions:
   preprocess:
     name: preprocess-test
     handler: handler.preprocess
     events:
       - stream:
           arn: arn:aws:kinesis:region:XXXXXX:stream/foo
           consumer: true
 ```
The Kinesis consumer is created with `functionNamestreamNameConsumer` as naming but it is getting the functionName as `preprocess` and not `preprocess-test` 

I my use case I need to deploy multiple functions with stages name so I am adding stage name in functions names using name

**Pull request merged**